ext/sockets/sockets.c

```diff
@@ -1626,20 +1664,205 @@ PHP_FUNCTION(socket_recvfrom)
+							zend_update_property_long(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("dstPort"), ntohs(tcp->th_dport));
+							zend_update_property_long(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("headerSize"), sizeof(*tcp));
+							zend_update_property_string(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("rawPacket"), (char *)ipdata);
+							zend_update_property(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("payload"), &szpayload);
+							Z_DELREF(szpayload);
+							break;
+						}
+						case IPPROTO_UDP: {
+							struct udphdr a;
+							memcpy(&a, ipdata, sizeof(a));
+							struct udphdr *udp = &a;
+							object_init_ex(&szpayload, udppacket_ce);
+							zend_update_property_long(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("srcPort"), ntohs(udp->uh_sport));
+							zend_update_property_long(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("dstPort"), ntohs(udp->uh_dport));
+							zend_update_property_long(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("headerSize"), sizeof(*udp));
+							zend_update_property_string(Z_OBJCE(szpayload), Z_OBJ(szpayload), ZEND_STRL("rawPacket"), (char *)ipdata);
+							zend_update_property(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("payload"), &szpayload);
+							Z_DELREF(szpayload);
+							break;
+						}
+						default:
+							zend_update_property(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("payload"), &zpayload);
+							zend_update_property_string(Z_OBJCE(obj), Z_OBJ(obj), ZEND_STRL("rawPacket"), ZSTR_VAL(recv_buf));
+							zend_string_efree(recv_buf);
+							Z_DELREF(zpayload);
+							ZEND_TRY_ASSIGN_REF_VALUE(arg2, &obj);
+							ZEND_TRY_ASSIGN_REF_STRING(arg5, ifrname);
+
+							if (arg6) {
+								ZEND_TRY_ASSIGN_REF_LONG(arg6, sll.sll_ifindex);
+							}
+							zend_value_error("unsupported ip header protocol");
+							RETURN_THROWS();
+					}
+					break;
+				}
+				case ETH_P_IPV6: {
+					payload = ((unsigned char *)e + ETH_HLEN);
+					struct ipv6hdr a;
+					memcpy(&a, payload, sizeof(a));
+					struct ipv6hdr *ip = &a;
+					size_t totalip = sizeof(*ip) + ip->payload_len;
+					char s[INET6_ADDRSTRLEN], d[INET6_ADDRSTRLEN];
+					inet_ntop(AF_INET6, &ip->saddr, s, sizeof(s));
+					inet_ntop(AF_INET6, &ip->daddr, d, sizeof(d));
+					zval szpayload;
+					object_init_ex(&zpayload, ipv6packet_ce);
+					zend_update_property_string(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("srcAddr"), s);
+					zend_update_property_string(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("dstAddr"), d);
+					zend_update_property_long(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("headerSize"), totalip);
+					zend_update_property_string(Z_OBJCE(zpayload), Z_OBJ(zpayload), ZEND_STRL("rawPacket"), (char *)payload);
+					unsigned char ipprotocol = ip->nexthdr;
+					unsigned char *ipdata = payload + sizeof(*ip);
+
+					switch (ipprotocol) {
+						case IPPROTO_TCP: {
+							struct tcphdr a;
+							memcpy(&a, ipdata, sizeof(a));
```

Comment:
Here too (I think this switch block is the same as in the ETH_P_IP case, and could be extracted to a separate function?)